include/swift/AST/AutoDiff.h

```diff
@@ -79,6 +79,123 @@ class AutoDiffParameter {
 };
 
 class AnyFunctionType;
+class Type;
+
+/// Identifies a subset of a function's parameters.
+///
+/// Works with AST-level function decls and types. Requires further lowering to
+/// work with SIL-level functions and types. (In particular, tuples must be
+/// exploded).
+class AutoDiffParameterIndices {
+  /// Bits corresponding to parameters in the set are "on", and bits
+  /// corresponding to parameters not in the set are "off".
+  ///
+  /// Normally, the bits correspond to the function's parameters in order. For
+  /// example,
+  ///
+  ///   Function type: (A, B, C) -> R
+  ///   Bits: [A][B][C]
+  ///
+  /// When `isMethod` is set, the bits correspond to the function's non-self
+  /// parameters in order, followed by the function's self parameter. For
+  /// example,
+  ///
+  ///   Function type: (Self) -> (A, B, C) -> R
+  ///   Bits: [A][B][C][Self]
+  ///
+  llvm::SmallBitVector indices;
+
+  /// Whether the function is a method.
+  ///
+  bool isMethod;
+
+  unsigned getNonSelfParamCount() const;
+
+  AutoDiffParameterIndices(unsigned indicesSize, bool isMethod)
+      : indices(indicesSize), isMethod(isMethod) {}
+
+public:
+  /// Allocates and initializes an empty `AutoDiffParameterIndices` for the
+  /// given `functionType`. `isMethod` specifies whether to treat the function
+  /// as a method.
+  static AutoDiffParameterIndices *create(ASTContext &C,
+                                          AnyFunctionType *functionType,
+                                          bool isMethod);
+
+  bool getIsMethod() const { return isMethod; }
+
+  /// Tests whether none of the parameters are in the set.
+  bool none() const { return indices.none(); }
+
+  /// Adds the indexed parameter to the set. When `isMethod` is not set, the
+  /// indices index into the first parameter list. For example,
+  ///
+  ///   functionType = (A, B, C) -> R
+  ///   paramIndex = 0
+  ///   ==> adds "A" to the set.
+  ///
+  /// When `isMethod` is set, the indices index into the first non-self
+  /// parameter list. For example,
+  ///
+  ///   functionType = (Self) -> (A, B, C) -> R
+  ///   paramIndex = 0
+  ///   ==> adds "A" to the set.
+  ///
+  void addNonSelfParam(unsigned paramIndex);
```

Comment:
How about `setNonSelfParameter`? `add` is a little weird for sets. I think we should use "parameter" everywhere because the type name, as well as existing AD code, uses "parameter", not "param".